onnx/shape_inference/implementation.h

```diff
@@ -0,0 +1,145 @@
+  }
+  std::unordered_map<std::string, const AttributeProto *> attributesByName_;
+  std::vector<const TypeProto_Tensor*> allInputTypes_;
+  std::vector<TypeProto_Tensor> allOutputTypes_;
+};
+
+ void mergeShapesAndTypes(const TypeProto_Tensor& inferredType, TypeProto_Tensor* existingType, const std::string& output) {
+}
+
+void InferShapes(ModelProto& m) {
+  std::unordered_map<std::string, int> opset_imports;
+  for (const auto& opset_import : m.opset_import()) {
+    opset_imports[opset_import.domain()] = static_cast<int>(opset_import.version());
+  }
+
+  auto* g = m.mutable_graph();
+
+  std::unordered_map<std::string, TypeProto_Tensor *> valueTypesByName;
+  for (auto& vi : *g->mutable_value_info()) {
+    valueTypesByName[vi.name()] = vi.mutable_type()->mutable_tensor_type();
+  }
+  for (auto& vi : *g->mutable_input()) {
+    valueTypesByName[vi.name()] = vi.mutable_type()->mutable_tensor_type();
+  }
+  for (auto& vi : *g->mutable_output()) {
+    valueTypesByName[vi.name()] = vi.mutable_type()->mutable_tensor_type();
+  }
+
+  for (const auto& n : g->node()) {
+    // Resolve domain for node
+    auto dit = opset_imports.find(n.domain());
+    if (dit == opset_imports.end()) {
+      continue;
+    }
+    auto domain_version = dit->second;
+
+    auto schema = OpSchemaRegistry::Schema(n.op_type(), domain_version, n.domain());
+    if (!schema) {
+      continue;
+    }
+
+    InferenceContextImpl ctx(n, valueTypesByName);
+
+    schema->GetShapeInferenceFunction()(ctx);
+
+    for (int i = 0; i < n.output_size(); ++i) {
+      const auto& output = n.output(i);
+      const auto& inferredType = *ctx.getOutputType(i);
+
+      // In this case, we have no new information, so don't bother
+      // to add a contentless ValueInfo.
+      if (inferredType.elem_type() == TensorProto::UNDEFINED &&
+          !inferredType.has_shape()) {
+        continue;
+      }
+
+      // If there is already a ValueInfo associated with this
+      // output, reuse it. Otherwise add a new one.
+      auto iter = valueTypesByName.find(output);
+      TypeProto_Tensor* existingType = nullptr;
+      if (iter != valueTypesByName.end()) {
+        existingType = iter->second;
+      } else {
+        auto vi = g->add_value_info();
+        vi->set_name(output);
+        existingType = vi->mutable_type()->mutable_tensor_type();
+      }
+
+      // Incorporate the inferred information.
+      if (inferredType.elem_type() != TensorProto::UNDEFINED) {
+        if (existingType->elem_type() != TensorProto::UNDEFINED &&
+            existingType->elem_type() != inferredType.elem_type()) {
+          throw std::runtime_error("inferred type differs from existing type");
+        } else {
+          existingType->set_elem_type(inferredType.elem_type());
+        }
+      }
+
+      for (int j = 0; j < inferredType.shape().dim_size(); ++j) {
+        const auto& inferredDim = inferredType.shape().dim(j);
+        if (existingType->mutable_shape()->dim_size() <= j) {
```

Comment:
This doesn't need to call "mutable"_shape(), in general we should be careful about calling it because it can change the semantics of whether the shape is not set or it's a scalar.